# Pathway-level analyses and gene networks

Work in progress.

Analyses to consider:

* [N-of-1-pathways](http://www.lussierlab.net/publications/N-of-1-pathways/)

>*N-of-1-pathways* is the 1st analytic framework and tool designed for personal transcriptome profiling and interpretation to predict clinical outcomes, analyze response to therapies, and monitor disease progression - one patient at a time ([1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4215042/), [2](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4765863/pdf/btv253.pdf)).

* ssGSEA

>A common approach to interpreting gene expression data is gene set enrichment analysis (GSEA) based on the functional annotation of the differentially expressed genes. This is useful for finding out if the differentially expressed genes are associated with a certain biological process or molecular function. ([1](https://www.ncbi.nlm.nih.gov/pubmed/19847166/), [2](https://www.ncbi.nlm.nih.gov/pubmed/16199517)).

* FAIME

>The Functional Analysis of Individual Microarray Expression (FAIME), is an approach to generate “personal mechanism signatures” of molecular pathways and functions from gene expression arrays. It computes mechanism scores using rank-weighted gene expression of an individual sample ([1](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002350)).

* [Pathway-Visualiser](https://github.com/algarnham/Pathway-Visualiser)
 
A shiny implementation to visualise enriched Gene Ontologies (GO), containing standardised annotation of gene products, identified by comparing the frequency of individual annotations in the gene list (e.g differentially expressed genes) with a reference list (all genes in the genome).

* Gene networks

>Display interactions between genes of interest and their primary neighbours using an interactive network, with nodes (genes) coloured according to the expression level (similar to the function implemented in [PIXdb](http://pixdb.org.uk/)). The networks are based on manually curated protein-protein interaction from [mentha](https://www.ncbi.nlm.nih.gov/pubmed/23900247) database.


## Table of contents

<!-- vim-markdown-toc GFM -->
* [Scripts summary](#scripts-summary)
* [Post-process, summarise and visualise sample data](#post-process-summarise-and-visualise-sample-data)
  * [Report content](#report-content)
  * [Example report](#example-report)
  * [To-do list](#to-do-list)


<!-- vim-markdown-toc -->
<br>


## Scripts summary

To be added

Script | Description | Packages
------------ | ------------ | ------------
... | ... | To be completed

<br />


## Post-process summarise and visualise sample data


**Script**: 

Argument | Description
------------ | ------------
... | To be completed
<br />

**Command line use example**:

```
```
<br>

Output file | Description
------------ | -----------
... | To be completed
<br />

### Report content

To be added

### Example report

To be added

```
```


### To-do list

* [Cumulative distribution function](https://en.wikipedia.org/wiki/Cumulative_distribution_function) (CDF) plots for selected genes to present their expression levels in the context of the overall expression distribution in investigated sample
* Link to [umccrise](https://github.com/umccr/umccrise) report to define genes to for the  CDF plots
